은행 창구 관리앱 [STEP 2] - d.o & joo

ohdair opened the pull request:
안녕하세요🙋‍♂️

이번 STEP2 에서는 `타입추상화`와 `Queue` 생성이 중점이였어요.! 

그래서 기능 구현보다 역할과 책임을 나누는 부분을 고민하며 구현하였습니다

## **Step 2 : 타입 구현 및 콘솔앱 구현**

### **타입 구현 - 은행, 고객**

- [x]  은행에는 n명의 은행원이 근무합니다
- [x]  은행에는 n명의 고객이 업무처리를 위해 대기합니다
    - [x]  고객의 대기열은 Step 1에서 구현한 Queue 타입을 활용합니다
- [x]  모든 고객의 업무가 끝나면 은행업무를 마감합니다
- [x]  업무를 마감할 때 "업무가 마감되었습니다. 오늘 업무를 처리한 고객은 총 XX명이며, 총 업무시간은 XX초니다."라고 출력합니다
- [x]  은행원은 고객의 업무를 처리합니다
- [x]  각 고객의 업무를 처리하는 데 걸리는 시간은 0.7초입니다
- [x]  은행원이 한 번에 처리할 수 있는 고객은 한 명입니다
- [x]  대기중인 고객의 업무 시작과 종료시 메세지 출력

### **콘솔앱 구현**

- [x]  Step 2의 은행에는 한 명의 은행원이 근무합니다.
- [x]  은행원은 한 번에 한 명의 고객의 업무를 처리할 수 있습니다.
- [x]  앱을 실행하면 두 개의 메뉴를 출력합니다.1 : 은행 개점2 : 종료
- [x]  사용자가 1을 입력하면 은행을 개점하고 10명~30명의 고객이 방문합니다. 10~30 사이의 임의의 수 만큼의 고객의 업무를 처리하면 은행문이 닫히고 다시 메뉴를 출력합니다.
- [x]  사용자가 2를 입력하면 프로그램을 종료합니다.

### **스텝 핵심 경험**

- Queue의 활용
- 타입 추상화 및 일반화

# 질문

1️⃣  타입을 쪼개고 생성하다 보니 파일이 많아졌는데요! 폴더링을 하면 구조가 깔끔해 질 것 같은데! 뷰나 뷰컨트롤러처럼 명확하게 나눠지지 않다보니 어떤 기준으로 잘라야 하는지 감이 잘 안오더라구요😭 라자냐는 어떤 방식으로 나누시는지 궁금해요❕

2️⃣ 매직넘버와 스트링을 관리하는 경우 `enum`으로 `static let` 혹은 인수가 필요한 문자열의 경우는 `static func`로 만들어서 사용했어요. 라자냐는 어떻게 관리를 하는지 궁금해요

ohdair committed:
feat: 프로그램 실행시 console에 두 개의 메뉴를 출력하는 함수 추가

altokeys committed:
feat: 은행업무 기능 추가

teller 구조체 추가

ohdair committed:
refactor: Bank와 BankManager로 분리

- 은행 업무시 메시지 출력을 BankMessage로 분리
- 대기하는 Customer 생성하는 메서드 분리

altokeys committed:
refactor: Bank 에 dequeue() 메서드 추가
- BankManager dequeue() 메서드 삭제

ohdair committed:
refactor: 은행 startBanking 메서드를 분리

- BankState로 은행 개점과 폐점의 상태로 구분
- 잘못된 입력에 대한 Error 메시지 출력
- execute에서 사용자의 입력을 받고 은행 업무 startBanking을 하도록 변경
- createCustomer 메서드를 visitedCustomers로 변경

altokeys committed:
refactor: BankError, BankMessage 파일 추가

ohdair committed:
refactor: 매직 스트링을 ConsoleMessage로 파일 분리

altokeys committed:
refactor: 변수명 변경 및 파일 구조 변경

altokeys committed:
refactor: 변수명 변경

wonhee009 commented on BankManagerConsoleApp/BankManagerConsoleApp/Model/BankMessage.swift:
```diff
@@ -0,0 +1,24 @@
+        case .endBanking(let customers, let takenTime):
+            print(ConsoleMessage.endBanking(customers: customers, takenTime: takenTime))
+        }
+    }
```
메서드 이름이 description인데 내부 동작은 출력까지 하고 있는게 조금 어색하게 느껴집니다.
보통은 `description`을 computed property로 정의해서 사용하는거 같고, 해당 메시지를 출력하는 역할을 Bank가 갖는게 맞지 않을까요?

doshkor replied:
`enum ConsoleMessage` 에서 메세지를 관리하고, `var description: String {}` 에서 메세지를 리턴합니다
`protocol ConsoleMessagable` 에서 `func printMessage(message: ConsoleMessage) {}` 를 구현하였습니다

Bank가 해당 프로토콜을 채택하여 메세지를 출력하게 구현하였습니다!

wonhee009 commented on BankManagerConsoleApp/BankManagerConsoleApp/Bank.swift:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    private func command() throws -> BankState {
+        guard let input = readLine() else {
+            return .close
+        }
```
input을 입력받지 못하면 은행이 닫히는거 같은데요. 사용자한테 다시 입력 받지 않아도 되는걸까요? 아니면 에러를 던지는 방법도 괜찮을거 같습니다.
(어쨌든 사용자가 2를 입력한 것은 아니기 때문에...)

ohdair replied:
터미널에서는 `ctrl + d`를 하게 되면 동작하고 있는 프로그램을 종료하는 명령을 처리 하기 위해서 guard문으로 readLine()의 입력이 되지 않으면서 종료를 하려는 명령을 받도록 처리 했어요

wonhee009 commented on BankManagerConsoleApp/BankManagerConsoleApp/Bank.swift:
```diff
@@ -0,0 +1,65 @@
+        BankMessage.startBanking.description()
+        do {
+            let command = try command()
+            guard command == .open else {
+                return
+            }
```
현재는 open이 아니면 Bank의 로직이 끝나는거 같은데요.
이후에 `enum BankState`가 추가될 경우도 고려해서 switch문을 이용해보는건 어떨까요?

doshkor replied:
말씀하신대로 enum case 추가시를 고려하여 swtich 가 적절한 것 같아 해당 구문 `switch`로 변경하였습니다

해당 커밋: d22daeb

wonhee009 commented on BankManagerConsoleApp/BankManagerConsoleApp/Bank.swift:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    private func startBanking() {
+        let visitedCustomers = visitedCustomers()
```
visitedCustomers() 내에서 큐에 customer들을 enqueue한 후에 은행 로직을 돌면서 customer를 dequeue하는거로 이해했습니다.
visitedCustomers라는 프로퍼티를 정의한게 은행이 끝날때 처리한 고객의 수를 위해서로 이해했습니다.

- visitedCustomers()라는 네이밍으로 봐서는 Customer 타입을 반환해줄거 같은데 고객의 수(Int)를 반환하는 부분이 어색한거 같습니다. 또한, visitedCustomers()는 은행에서 처리해야하는 고객을 큐에 넣어주는 역할인거 같은데 해당 역할에서 고객의 수를 반환하는 부분 또한 어색한거 같습니다.
- 고객의 수를 들고 있지 않고 dequeue 과정을 통해 고객의 수를 셀 수 있지 않을까요?

ohdair replied:
로직상 업무 처리 후, 인원은 확인하는 게 맞다고 생각되어서 enqueue하는 메서드로 변경했어요

wonhee009 commented on BankManagerConsoleApp/BankManagerConsoleApp/Bank.swift:
```diff
@@ -0,0 +1,65 @@
+        return command
+    }
+
+    private func startBanking() {
```
해당 메서드에서 굉장히 많은 일을 하고 있는거 같습니다.
고객 생성, 운영 시간 기록, 마감 메시지 출력
역할을 나눠보는건 어떤가요?

ohdair replied:
많은 일을 줄이기 위해 `Sequence`, `Iteratorable`를 사용하여 고차함수를 사용할 수 있도록 했어요.
1. 고객 생성하는 메서드 생성
2. 영업 시간을 기록을 위해 extension으로 분리
3. 고객 업무 처리하는 메서드 생성

고객 업무 처리 메서드 
1. 대기줄의 고객 dequeue
2. 고객 업무 처리
3. 하루 방문 고객 업데이트
```swift
private func serveCustomers() {
    Bank.customersPerDay = queue.reduce(0) { customersPerDay, customer in
        queue.dequeue()
        teller.assist(customer)
        return customersPerDay + 1
    }
}
```

wonhee009 reviewed (changes requested):
수고하셨습니다 :)
코멘트 남겨두었으니 확인 부탁드려요.

- 프로젝트마다 달라서 이렇게 한다고 말씀은 못 드릴거 같습니다만 현재 보내주신 프로젝트를 제가 폴더링 한다면.... LinkedList, Queue 정도를 Model로 둘거 같습니다. 나머지는 Constant한 성격의 파일들끼리 폴더링할거 같습니다. 사실 이게 정답이 없고, 최대한 비슷한 역할들끼리 모아둘거 같아요. 이런 부분들은 회사/프로젝트마다 갖고 가는 방식이 달라서 정답은 없습니다~
- 저도 보통 그렇게 관리합니다만, 프로젝트 전반적으로 사용하는 부분이 아니라면 `private enum`으로 사용하는 파일 내에 정의해서 사용하는 편입니다.

ohdair committed:
refactor: ConsoleMessagable 프로토콜 추가

- ConsoleMessage를 연산 프로퍼티로 변경
- ConsoleMessage 타입을 출력할 수 있는 프로토콜을 생성
- printMessage 메서드로 ConsoleMessage 타입을 출력
- 역할별 폴더 구분

altokeys committed:
refactor: 사용자입력시 수행 로직 변경
- guard문에서 switch문으로 변경

ohdair committed:
refactor: startBanking 메서드의 역할을 분리하기 위해 extention 추가

- 현재의 시간을 Double로 가져오는 now() 메서드 추가
- 특정 시간부터 현재 시간까지의 측정된 값을 가져오는 takenTime(from:) 메서드 추가
- Double 타입에서 소숫점 첫째짜리 까지 버림을 하는 floor() 메서드 추가
- 방문한 고객들을 확인하는 customersPerDay 추가

wonhee009 commented on BankManagerConsoleApp/BankManagerConsoleApp/Bank/Bank.swift:
```diff
@@ -0,0 +1,87 @@
+    }
+}
+
+extension Date {
```
```suggestion
private extension Date {
```

wonhee009 commented on BankManagerConsoleApp/BankManagerConsoleApp/Bank/Bank.swift:
```diff
@@ -0,0 +1,87 @@
+
+    private func enqueueCustomers() {
+        let customers = Int.random(in: Constants.rangeOfCustomers)
+        for customer in 1...customers {
+            queue.enqueue(customer)
+        }
```
`foreach` 사용할 수 있지 않나요?.?

wonhee009 reviewed (approved):
수고하셧습니다 :)
남겨둔 코멘트는 이후 스텝에 반영 부탁드려요.
